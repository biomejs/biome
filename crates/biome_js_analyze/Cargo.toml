[package]
authors.workspace    = true
edition.workspace    = true
license.workspace    = true
name                 = "biome_js_analyze"
repository.workspace = true
version              = "0.2.0"

[dependencies]
biome_analyze          = { workspace = true }
biome_aria             = { workspace = true }
biome_console          = { workspace = true }
biome_control_flow     = { workspace = true }
biome_deserialize      = { workspace = true }
biome_diagnostics      = { workspace = true }
biome_js_factory       = { workspace = true }
biome_js_semantic      = { workspace = true }
biome_js_syntax        = { workspace = true }
biome_js_unicode_table = { workspace = true }
biome_json_factory     = { workspace = true }
biome_json_syntax      = { workspace = true }
biome_rowan            = { workspace = true }
bpaf.workspace         = true
lazy_static            = { workspace = true }
natord                 = "1.0.9"
roaring                = "0.10.1"
rustc-hash             = { workspace = true }
schemars               = { workspace = true, optional = true }
serde                  = { workspace = true, features = ["derive"] }
serde_json             = { workspace = true }
smallvec               = { workspace = true }

[dev-dependencies]
biome_js_parser  = { workspace = true, features = ["tests"] }
biome_test_utils = { workspace = true }
biome_text_edit  = { workspace = true }
countme          = { workspace = true, features = ["enable"] }
insta            = { workspace = true, features = ["glob"] }
tests_macros     = { workspace = true }

[features]
schema = ["schemars", "biome_deserialize/schema"]
