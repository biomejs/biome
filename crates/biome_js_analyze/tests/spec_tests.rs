use biome_analyze::{
    AnalysisFilter, AnalyzerAction, AnalyzerPluginSlice, ControlFlow, Never, RuleFilter,
};
use biome_diagnostics::advice::CodeSuggestionAdvice;
use biome_fs::OsFileSystem;
use biome_js_analyze::JsAnalyzerServices;
use biome_js_parser::{parse, JsParserOptions};
use biome_js_syntax::{JsFileSource, JsLanguage, ModuleKind};
use biome_package::PackageType;
use biome_plugin_loader::AnalyzerGritPlugin;
use biome_rowan::AstNode;
use biome_test_utils::{
    assert_errors_are_absent, code_fix_to_string, create_analyzer_options,
    dependency_graph_for_test_file, diagnostic_to_string, has_bogus_nodes_or_empty_slots,
    parse_test_path, project_layout_with_node_manifest, register_leak_checker, scripts_from_json,
    write_analyzer_snapshot, CheckActionType,
};
use camino::{Utf8Component, Utf8Path};
use std::ops::Deref;
use std::sync::Arc;
use std::{fs::read_to_string, slice};

tests_macros::gen_tests! {"tests/specs/**/*.{cjs,cts,js,jsx,tsx,ts,json,jsonc,svelte}", crate::run_test, "module"}
tests_macros::gen_tests! {"tests/suppression/**/*.{cjs,cts,js,jsx,tsx,ts,json,jsonc,svelte}", crate::run_suppression_test, "module"}
tests_macros::gen_tests! {"tests/plugin/*.grit", crate::run_plugin_test, "module"}

fn run_test(input: &'static str, _: &str, _: &str, _: &str) {
    register_leak_checker();

    let input_file = Utf8Path::new(input);
    let file_name = input_file.file_name().unwrap();

    let (group, rule) = parse_test_path(input_file);
    if rule == "specs" || rule == "suppression" {
        panic!("the test file must be placed in the {rule}/<group-name>/<rule-name>/ directory");
    }
    if group == "specs" || group == "suppression" {
        panic!("the test file must be placed in the {group}/{rule}/<rule-name>/ directory");
    }
    if biome_js_analyze::METADATA
        .deref()
        .find_rule(group, rule)
        .is_none()
    {
        panic!("could not find rule {group}/{rule}");
    }

    let rule_filter = RuleFilter::Rule(group, rule);
    let filter = AnalysisFilter {
        enabled_rules: Some(slice::from_ref(&rule_filter)),
        ..AnalysisFilter::default()
    };

    let mut snapshot = String::new();
    let extension = input_file.extension().unwrap_or_default();

    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {input_file:?}: {err:?}"));
    let quantity_diagnostics = if let Some(scripts) = scripts_from_json(extension, &input_code) {
        for script in scripts {
            analyze_and_snap(
                &mut snapshot,
                &script,
                JsFileSource::js_script(),
                filter,
                file_name,
                input_file,
                CheckActionType::Lint,
                JsParserOptions::default(),
                &[],
            );
        }

        0
    } else {
        let Ok(source_type) = input_file.try_into() else {
            return;
        };
        analyze_and_snap(
            &mut snapshot,
            &input_code,
            source_type,
            filter,
            file_name,
            input_file,
            CheckActionType::Lint,
            JsParserOptions::default(),
            &[],
        )
    };

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });

    if input_code.contains("/* should not generate diagnostics */") && quantity_diagnostics > 0 {
        panic!("This test should not generate diagnostics");
    }
}

#[expect(clippy::too_many_arguments)]
pub(crate) fn analyze_and_snap(
    snapshot: &mut String,
    input_code: &str,
    mut source_type: JsFileSource,
    filter: AnalysisFilter,
    file_name: &str,
    input_file: &Utf8Path,
    check_action_type: CheckActionType,
    parser_options: JsParserOptions,
    plugins: AnalyzerPluginSlice,
) -> usize {
    let mut diagnostics = Vec::new();
    let mut code_fixes = Vec::new();
    let project_layout = project_layout_with_node_manifest(input_file, &mut diagnostics);

    if let Some((_, manifest)) = project_layout.get_node_manifest_for_path(input_file) {
        if manifest.r#type == Some(PackageType::CommonJs) &&
            // At the moment we treat JS and JSX at the same way
            (source_type.file_extension() == "js" || source_type.file_extension() == "jsx" )
        {
            source_type.set_module_kind(ModuleKind::Script)
        }
    }

    let parsed = parse(input_code, source_type, parser_options.clone());
    let root = parsed.tree();

    let options = create_analyzer_options(input_file, &mut diagnostics);

    // FIXME: We probably want to enable it for all rules? Right now it seems to
    //        trigger a leak panic...
    let dependency_graph = if input_file
        .components()
        .any(|component| component == Utf8Component::Normal("noImportCycles"))
    {
        dependency_graph_for_test_file(input_file, &project_layout)
    } else {
        Default::default()
    };

    let services = JsAnalyzerServices::from((dependency_graph, project_layout, source_type));

    let (_, errors) =
        biome_js_analyze::analyze(&root, filter, &options, plugins, services, |event| {
            if let Some(mut diag) = event.diagnostic() {
                for action in event.actions() {
                    if check_action_type.is_suppression() {
                        if action.is_suppression() {
                            check_code_action(
                                input_file,
                                input_code,
                                source_type,
                                &action,
                                parser_options.clone(),
                            );
                            diag = diag.add_code_suggestion(CodeSuggestionAdvice::from(action));
                        }
                    } else if !action.is_suppression() {
                        check_code_action(
                            input_file,
                            input_code,
                            source_type,
                            &action,
                            parser_options.clone(),
                        );
                        diag = diag.add_code_suggestion(CodeSuggestionAdvice::from(action));
                    }
                }

                diagnostics.push(diagnostic_to_string(file_name, input_code, diag.into()));
                return ControlFlow::Continue(());
            }

            for action in event.actions() {
                if check_action_type.is_suppression() {
                    if action.category.matches("quickfix.suppressRule") {
                        check_code_action(
                            input_file,
                            input_code,
                            source_type,
                            &action,
                            parser_options.clone(),
                        );
                        code_fixes.push(code_fix_to_string(input_code, action));
                    }
                } else if !action.category.matches("quickfix.suppressRule") {
                    check_code_action(
                        input_file,
                        input_code,
                        source_type,
                        &action,
                        parser_options.clone(),
                    );
                    code_fixes.push(code_fix_to_string(input_code, action));
                }
            }

            ControlFlow::<Never>::Continue(())
        });

    for error in errors {
        diagnostics.push(diagnostic_to_string(file_name, input_code, error));
    }

    write_analyzer_snapshot(
        snapshot,
        input_code,
        diagnostics.as_slice(),
        code_fixes.as_slice(),
        source_type.file_extension(),
    );

    // FIXME: I wish we could do this more generically, but we cannot do this
    //        for all tests, since it would cause many incorrect replacements.
    //        Maybe there's a regular expression that could work, but it feels
    //        flimsy too...
    if input_file
        .components()
        .any(|component| component == Utf8Component::Normal("noImportCycles"))
    {
        // Normalize Windows paths.
        *snapshot = snapshot.replace('\\', "/");
    }

    diagnostics.len()
}

fn check_code_action(
    path: &Utf8Path,
    source: &str,
    source_type: JsFileSource,
    action: &AnalyzerAction<JsLanguage>,
    options: JsParserOptions,
) {
    let (new_tree, text_edit) = match action
        .mutation
        .clone()
        .commit_with_text_range_and_edit(true)
    {
        (new_tree, Some((_, text_edit))) => (new_tree, text_edit),
        (new_tree, None) => (new_tree, Default::default()),
    };

    let output = text_edit.new_string(source);

    // Checks that applying the text edits returned by the BatchMutation
    // returns the same code as printing the modified syntax tree
    assert_eq!(
        new_tree.to_string(),
        output,
        "Code action and syntax tree differ"
    );

    if has_bogus_nodes_or_empty_slots(&new_tree) {
        panic!("modified tree has bogus nodes or empty slots:\n{new_tree:#?} \n\n {new_tree}")
    }

    // Checks the returned tree contains no missing children node
    if format!("{new_tree:?}").contains("missing (required)") {
        panic!("modified tree has missing children:\n{new_tree:#?}")
    }

    // Re-parse the modified code and panic if the resulting tree has syntax errors
    let re_parse = parse(&output, source_type, options);
    assert_errors_are_absent(re_parse.tree().syntax(), re_parse.diagnostics(), path);
}

pub(crate) fn run_suppression_test(input: &'static str, _: &str, _: &str, _: &str) {
    register_leak_checker();

    let input_file = Utf8Path::new(input);
    let file_name = input_file.file_name().unwrap();
    let source_type = match input_file.extension() {
        Some("js" | "mjs" | "jsx") => JsFileSource::jsx(),
        Some("cjs") => JsFileSource::js_script(),
        Some("ts") => JsFileSource::ts(),
        Some("mts" | "cts") => JsFileSource::ts_restricted(),
        Some("tsx") => JsFileSource::tsx(),
        _ => {
            panic!("Unknown file extension: {:?}", input_file.extension());
        }
    };
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {input_file:?}: {err:?}"));

    let (group, rule) = parse_test_path(input_file);

    let rule_filter = RuleFilter::Rule(group, rule);
    let filter = AnalysisFilter {
        enabled_rules: Some(slice::from_ref(&rule_filter)),
        ..AnalysisFilter::default()
    };

    let mut snapshot = String::new();
    analyze_and_snap(
        &mut snapshot,
        &input_code,
        source_type,
        filter,
        file_name,
        input_file,
        CheckActionType::Suppression,
        JsParserOptions::default(),
        &[],
    );

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });
}

fn run_plugin_test(input: &'static str, _: &str, _: &str, _: &str) {
    register_leak_checker();

    let plugin_path = Utf8Path::new(input);
    let file_name = plugin_path.file_name().unwrap();
    let input_path = plugin_path.with_extension("js");

    let plugin = match AnalyzerGritPlugin::load(
        &OsFileSystem::new(plugin_path.to_owned()),
        Utf8Path::new(plugin_path),
    ) {
        Ok(plugin) => plugin,
        Err(err) => panic!("Cannot load plugin: {err:?}"),
    };

    let filter = AnalysisFilter {
        enabled_rules: Some(&[]),
        ..AnalysisFilter::default()
    };

    let mut snapshot = String::new();

    let input_code = read_to_string(&input_path)
        .unwrap_or_else(|err| panic!("failed to read {input_path:?}: {err:?}"));
    let Ok(source_type) = input_path.as_path().try_into() else {
        return;
    };
    analyze_and_snap(
        &mut snapshot,
        &input_code,
        source_type,
        filter,
        file_name,
        &input_path,
        CheckActionType::Lint,
        JsParserOptions::default(),
        &[Arc::new(Box::new(plugin))],
    );

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => plugin_path.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });
}
