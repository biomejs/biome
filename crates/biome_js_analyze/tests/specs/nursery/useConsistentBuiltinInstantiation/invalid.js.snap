---
source: crates/biome_js_analyze/tests/spec_tests.rs
expression: invalid.js
---
# Input
```jsx
Object()
Array()
ArrayBuffer()
BigInt64Array()
BigUint64Array()
DataView()
Date()
Error()
Float32Array()
Float64Array()
Function()
Int8Array()
Int16Array()
Int32Array()
Map()
WeakMap()
Set()
WeakSet()
Promise()
RegExp()
Uint8Array()
Uint16Array()
Uint32Array()
Uint8ClampedArray()
SharedArrayBuffer()
Proxy()
WeakRef()
FinalizationRegistry()
window.Object({})
globalThis.Object()
function foo() {
    return /** Start */ globalThis.Object({ foo: 'bar' }) /** End */
}

new Boolean()
new BigInt()
new Number()
new String()
new Symbol()
new window.String(123)
new globalThis.String()
function foo() {
    return /** Start */ new globalThis.String("foo") /** End */
}
```

# Diagnostics
```
invalid.js:1:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Object() instead of Object().
  
  > 1 │ Object()
      │ ^^^^^^^^
    2 │ Array()
    3 │ ArrayBuffer()
  
  i Unsafe fix: Add new keyword.
  
    1 │ new·Object()
      │ ++++        

```

```
invalid.js:2:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Array() instead of Array().
  
    1 │ Object()
  > 2 │ Array()
      │ ^^^^^^^
    3 │ ArrayBuffer()
    4 │ BigInt64Array()
  
  i Unsafe fix: Add new keyword.
  
    2 │ new·Array()
      │ ++++       

```

```
invalid.js:3:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new ArrayBuffer() instead of ArrayBuffer().
  
    1 │ Object()
    2 │ Array()
  > 3 │ ArrayBuffer()
      │ ^^^^^^^^^^^^^
    4 │ BigInt64Array()
    5 │ BigUint64Array()
  
  i Unsafe fix: Add new keyword.
  
    3 │ new·ArrayBuffer()
      │ ++++             

```

```
invalid.js:4:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new BigInt64Array() instead of BigInt64Array().
  
    2 │ Array()
    3 │ ArrayBuffer()
  > 4 │ BigInt64Array()
      │ ^^^^^^^^^^^^^^^
    5 │ BigUint64Array()
    6 │ DataView()
  
  i Unsafe fix: Add new keyword.
  
    4 │ new·BigInt64Array()
      │ ++++               

```

```
invalid.js:5:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new BigUint64Array() instead of BigUint64Array().
  
    3 │ ArrayBuffer()
    4 │ BigInt64Array()
  > 5 │ BigUint64Array()
      │ ^^^^^^^^^^^^^^^^
    6 │ DataView()
    7 │ Date()
  
  i Unsafe fix: Add new keyword.
  
    5 │ new·BigUint64Array()
      │ ++++                

```

```
invalid.js:6:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new DataView() instead of DataView().
  
    4 │ BigInt64Array()
    5 │ BigUint64Array()
  > 6 │ DataView()
      │ ^^^^^^^^^^
    7 │ Date()
    8 │ Error()
  
  i Unsafe fix: Add new keyword.
  
    6 │ new·DataView()
      │ ++++          

```

```
invalid.js:7:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Date() instead of Date().
  
    5 │ BigUint64Array()
    6 │ DataView()
  > 7 │ Date()
      │ ^^^^^^
    8 │ Error()
    9 │ Float32Array()
  
  i Unsafe fix: Add new keyword.
  
    7 │ new·Date()
      │ ++++      

```

```
invalid.js:8:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Error() instead of Error().
  
     6 │ DataView()
     7 │ Date()
   > 8 │ Error()
       │ ^^^^^^^
     9 │ Float32Array()
    10 │ Float64Array()
  
  i Unsafe fix: Add new keyword.
  
    8 │ new·Error()
      │ ++++       

```

```
invalid.js:9:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Float32Array() instead of Float32Array().
  
     7 │ Date()
     8 │ Error()
   > 9 │ Float32Array()
       │ ^^^^^^^^^^^^^^
    10 │ Float64Array()
    11 │ Function()
  
  i Unsafe fix: Add new keyword.
  
    9 │ new·Float32Array()
      │ ++++              

```

```
invalid.js:10:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Float64Array() instead of Float64Array().
  
     8 │ Error()
     9 │ Float32Array()
  > 10 │ Float64Array()
       │ ^^^^^^^^^^^^^^
    11 │ Function()
    12 │ Int8Array()
  
  i Unsafe fix: Add new keyword.
  
    10 │ new·Float64Array()
       │ ++++              

```

```
invalid.js:11:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Function() instead of Function().
  
     9 │ Float32Array()
    10 │ Float64Array()
  > 11 │ Function()
       │ ^^^^^^^^^^
    12 │ Int8Array()
    13 │ Int16Array()
  
  i Unsafe fix: Add new keyword.
  
    11 │ new·Function()
       │ ++++          

```

```
invalid.js:12:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Int8Array() instead of Int8Array().
  
    10 │ Float64Array()
    11 │ Function()
  > 12 │ Int8Array()
       │ ^^^^^^^^^^^
    13 │ Int16Array()
    14 │ Int32Array()
  
  i Unsafe fix: Add new keyword.
  
    12 │ new·Int8Array()
       │ ++++           

```

```
invalid.js:13:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Int16Array() instead of Int16Array().
  
    11 │ Function()
    12 │ Int8Array()
  > 13 │ Int16Array()
       │ ^^^^^^^^^^^^
    14 │ Int32Array()
    15 │ Map()
  
  i Unsafe fix: Add new keyword.
  
    13 │ new·Int16Array()
       │ ++++            

```

```
invalid.js:14:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Int32Array() instead of Int32Array().
  
    12 │ Int8Array()
    13 │ Int16Array()
  > 14 │ Int32Array()
       │ ^^^^^^^^^^^^
    15 │ Map()
    16 │ WeakMap()
  
  i Unsafe fix: Add new keyword.
  
    14 │ new·Int32Array()
       │ ++++            

```

```
invalid.js:15:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Map() instead of Map().
  
    13 │ Int16Array()
    14 │ Int32Array()
  > 15 │ Map()
       │ ^^^^^
    16 │ WeakMap()
    17 │ Set()
  
  i Unsafe fix: Add new keyword.
  
    15 │ new·Map()
       │ ++++     

```

```
invalid.js:16:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new WeakMap() instead of WeakMap().
  
    14 │ Int32Array()
    15 │ Map()
  > 16 │ WeakMap()
       │ ^^^^^^^^^
    17 │ Set()
    18 │ WeakSet()
  
  i Unsafe fix: Add new keyword.
  
    16 │ new·WeakMap()
       │ ++++         

```

```
invalid.js:17:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Set() instead of Set().
  
    15 │ Map()
    16 │ WeakMap()
  > 17 │ Set()
       │ ^^^^^
    18 │ WeakSet()
    19 │ Promise()
  
  i Unsafe fix: Add new keyword.
  
    17 │ new·Set()
       │ ++++     

```

```
invalid.js:18:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new WeakSet() instead of WeakSet().
  
    16 │ WeakMap()
    17 │ Set()
  > 18 │ WeakSet()
       │ ^^^^^^^^^
    19 │ Promise()
    20 │ RegExp()
  
  i Unsafe fix: Add new keyword.
  
    18 │ new·WeakSet()
       │ ++++         

```

```
invalid.js:19:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Promise() instead of Promise().
  
    17 │ Set()
    18 │ WeakSet()
  > 19 │ Promise()
       │ ^^^^^^^^^
    20 │ RegExp()
    21 │ Uint8Array()
  
  i Unsafe fix: Add new keyword.
  
    19 │ new·Promise()
       │ ++++         

```

```
invalid.js:20:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new RegExp() instead of RegExp().
  
    18 │ WeakSet()
    19 │ Promise()
  > 20 │ RegExp()
       │ ^^^^^^^^
    21 │ Uint8Array()
    22 │ Uint16Array()
  
  i Unsafe fix: Add new keyword.
  
    20 │ new·RegExp()
       │ ++++        

```

```
invalid.js:21:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Uint8Array() instead of Uint8Array().
  
    19 │ Promise()
    20 │ RegExp()
  > 21 │ Uint8Array()
       │ ^^^^^^^^^^^^
    22 │ Uint16Array()
    23 │ Uint32Array()
  
  i Unsafe fix: Add new keyword.
  
    21 │ new·Uint8Array()
       │ ++++            

```

```
invalid.js:22:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Uint16Array() instead of Uint16Array().
  
    20 │ RegExp()
    21 │ Uint8Array()
  > 22 │ Uint16Array()
       │ ^^^^^^^^^^^^^
    23 │ Uint32Array()
    24 │ Uint8ClampedArray()
  
  i Unsafe fix: Add new keyword.
  
    22 │ new·Uint16Array()
       │ ++++             

```

```
invalid.js:23:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Uint32Array() instead of Uint32Array().
  
    21 │ Uint8Array()
    22 │ Uint16Array()
  > 23 │ Uint32Array()
       │ ^^^^^^^^^^^^^
    24 │ Uint8ClampedArray()
    25 │ SharedArrayBuffer()
  
  i Unsafe fix: Add new keyword.
  
    23 │ new·Uint32Array()
       │ ++++             

```

```
invalid.js:24:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Uint8ClampedArray() instead of Uint8ClampedArray().
  
    22 │ Uint16Array()
    23 │ Uint32Array()
  > 24 │ Uint8ClampedArray()
       │ ^^^^^^^^^^^^^^^^^^^
    25 │ SharedArrayBuffer()
    26 │ Proxy()
  
  i Unsafe fix: Add new keyword.
  
    24 │ new·Uint8ClampedArray()
       │ ++++                   

```

```
invalid.js:25:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new SharedArrayBuffer() instead of SharedArrayBuffer().
  
    23 │ Uint32Array()
    24 │ Uint8ClampedArray()
  > 25 │ SharedArrayBuffer()
       │ ^^^^^^^^^^^^^^^^^^^
    26 │ Proxy()
    27 │ WeakRef()
  
  i Unsafe fix: Add new keyword.
  
    25 │ new·SharedArrayBuffer()
       │ ++++                   

```

```
invalid.js:26:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Proxy() instead of Proxy().
  
    24 │ Uint8ClampedArray()
    25 │ SharedArrayBuffer()
  > 26 │ Proxy()
       │ ^^^^^^^
    27 │ WeakRef()
    28 │ FinalizationRegistry()
  
  i Unsafe fix: Add new keyword.
  
    26 │ new·Proxy()
       │ ++++       

```

```
invalid.js:27:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new WeakRef() instead of WeakRef().
  
    25 │ SharedArrayBuffer()
    26 │ Proxy()
  > 27 │ WeakRef()
       │ ^^^^^^^^^
    28 │ FinalizationRegistry()
    29 │ window.Object({})
  
  i Unsafe fix: Add new keyword.
  
    27 │ new·WeakRef()
       │ ++++         

```

```
invalid.js:28:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new FinalizationRegistry() instead of FinalizationRegistry().
  
    26 │ Proxy()
    27 │ WeakRef()
  > 28 │ FinalizationRegistry()
       │ ^^^^^^^^^^^^^^^^^^^^^^
    29 │ window.Object({})
    30 │ globalThis.Object()
  
  i Unsafe fix: Add new keyword.
  
    28 │ new·FinalizationRegistry()
       │ ++++                      

```

```
invalid.js:29:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Object() instead of Object().
  
    27 │ WeakRef()
    28 │ FinalizationRegistry()
  > 29 │ window.Object({})
       │ ^^^^^^^^^^^^^^^^^
    30 │ globalThis.Object()
    31 │ function foo() {
  
  i Unsafe fix: Add new keyword.
  
    29 │ new·window.Object({})
       │ ++++                 

```

```
invalid.js:30:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Object() instead of Object().
  
    28 │ FinalizationRegistry()
    29 │ window.Object({})
  > 30 │ globalThis.Object()
       │ ^^^^^^^^^^^^^^^^^^^
    31 │ function foo() {
    32 │     return /** Start */ globalThis.Object({ foo: 'bar' }) /** End */
  
  i Unsafe fix: Add new keyword.
  
    30 │ new·globalThis.Object()
       │ ++++                   

```

```
invalid.js:32:25 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use new Object() instead of Object().
  
    30 │ globalThis.Object()
    31 │ function foo() {
  > 32 │     return /** Start */ globalThis.Object({ foo: 'bar' }) /** End */
       │                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    33 │ }
    34 │ 
  
  i Unsafe fix: Add new keyword.
  
    32 │ ····return·/**·Start·*/·new·globalThis.Object({·foo:·'bar'·})·/**·End·*/
       │                         ++++                                            

```

```
invalid.js:35:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use Boolean() instead of new Boolean().
  
    33 │ }
    34 │ 
  > 35 │ new Boolean()
       │ ^^^^^^^^^^^^^
    36 │ new BigInt()
    37 │ new Number()
  
  i Unsafe fix: Remove new keyword.
  
    35 │ new·Boolean()
       │ ----         

```

```
invalid.js:36:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use BigInt() instead of new BigInt().
  
    35 │ new Boolean()
  > 36 │ new BigInt()
       │ ^^^^^^^^^^^^
    37 │ new Number()
    38 │ new String()
  
  i Unsafe fix: Remove new keyword.
  
    36 │ new·BigInt()
       │ ----        

```

```
invalid.js:37:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use Number() instead of new Number().
  
    35 │ new Boolean()
    36 │ new BigInt()
  > 37 │ new Number()
       │ ^^^^^^^^^^^^
    38 │ new String()
    39 │ new Symbol()
  
  i Unsafe fix: Remove new keyword.
  
    37 │ new·Number()
       │ ----        

```

```
invalid.js:38:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use String() instead of new String().
  
    36 │ new BigInt()
    37 │ new Number()
  > 38 │ new String()
       │ ^^^^^^^^^^^^
    39 │ new Symbol()
    40 │ new window.String(123)
  
  i Unsafe fix: Remove new keyword.
  
    38 │ new·String()
       │ ----        

```

```
invalid.js:39:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use Symbol() instead of new Symbol().
  
    37 │ new Number()
    38 │ new String()
  > 39 │ new Symbol()
       │ ^^^^^^^^^^^^
    40 │ new window.String(123)
    41 │ new globalThis.String()
  
  i Unsafe fix: Remove new keyword.
  
    39 │ new·Symbol()
       │ ----        

```

```
invalid.js:40:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use String() instead of new String().
  
    38 │ new String()
    39 │ new Symbol()
  > 40 │ new window.String(123)
       │ ^^^^^^^^^^^^^^^^^^^^^^
    41 │ new globalThis.String()
    42 │ function foo() {
  
  i Unsafe fix: Remove new keyword.
  
    40 │ new·window.String(123)
       │ ----                  

```

```
invalid.js:41:1 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use String() instead of new String().
  
    39 │ new Symbol()
    40 │ new window.String(123)
  > 41 │ new globalThis.String()
       │ ^^^^^^^^^^^^^^^^^^^^^^^
    42 │ function foo() {
    43 │     return /** Start */ new globalThis.String("foo") /** End */
  
  i Unsafe fix: Remove new keyword.
  
    41 │ new·globalThis.String()
       │ ----                   

```

```
invalid.js:43:25 lint/nursery/useConsistentBuiltinInstatiation  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Use String() instead of new String().
  
    41 │ new globalThis.String()
    42 │ function foo() {
  > 43 │     return /** Start */ new globalThis.String("foo") /** End */
       │                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    44 │ }
  
  i Unsafe fix: Remove new keyword.
  
    43 │ ····return·/**·Start·*/·new·globalThis.String("foo")·/**·End·*/
       │                         ----                                   

```
